completion/services.py

```diff
@@ -117,8 +117,14 @@ def vertical_is_complete(self, item):
         if not self.completion_tracking_enabled():
             return None
 
+        optional_vertical = getattr(item, "optional_content", False)
+
         # this is temporary local logic and will be removed when the whole course tree is included in completion
-        child_locations = [child.scope_ids.usage_id for child in self.get_completable_children(item)]
+        child_locations = [
+            child.scope_ids.usage_id for child in self.get_completable_children(item)
+            # for non-optional verticals, only include non-optional children
+            if optional_vertical or not getattr(child, "optional_content", False)
```

Comment:
@DanielVZ96, shouldn't this be something like `not optional_vertical and not getattr(child, "optional_content", False)`? We should add a test for this. We could also use a variable like `required_completion = getattr(item, "optional_content", False)` to avoid too many negations that make this condition hard to read.

Reply:
There are only four combinations:
1. Optional Vertical and Optional Child -> Include Child
2. Optional Vertical and Required Child -> Include Child (shouldn't happen but this is how we can handle it gracefully)
3. Required Vertical and Required Child -> Include Child
4. Required Vertical and Optional Child -> Exclude Child

This means that the only case in which we want to exclude the child
is when the parent is required and the child isn't.

This can be achieved with either of the following logic (nots turns ors into ands)
```
        a) optional_vertical or required_child  # can be read as "if it's an optional vertical include all, if it's a required vertical only accept required children"
        b) not (required_vertical and optional_child)  # can be read as "exclude the 4th case" 
```

Your's (`not optional_vertical and not getattr(child, "optional_content", False)`) can be paraphrased as `required_vertical and required_child`, and would only return true (thus include) the third case.

I'm refactoring the code to include this explanation and make it easier to test, since the vertical_is_complete is really hard to test without having edx-platform code available.